zenith_utils/src/postgres_backend.rs

```diff
@@ -74,16 +84,42 @@ pub fn query_from_cstring(query_string: Bytes) -> Vec<u8> {
     query_string
 }
 
+lazy_static! {
+    static ref POSTGRES_BACKEND_MSG_COUNT: IntCounterVec = register_int_counter_vec!(
+        "postgres_backend_message_count",
+        "Number of PostgreSQL protocol messages sent & received",
+        &["pg_backend_name", "pg_backend_msg_direction"]
+    )
+    .expect("failed to define metric");
+    static ref POSTGRES_BACKEND_MSG_TIME: HistogramVec = register_histogram_vec!(
+        "postgres_backend_process_message_time",
+        "Amount of time spent",
+        &["pg_backend_name", "pg_backend_msg_type"]
+    )
+    .expect("failed to define a metric");
+}
```

Comment:
IMO this counter doesn't tell us a lot -- let's have the same kind of logic in SMGR_QUERY_COUNTER, namely track time distribution and parametrize it with smgr query type.

Reply:
And one more thing. Default histogram bins are following [0.005,0.01,0.025,0.05,0.1,0.25,0.5,1,2.5,5,10,+Inf] and if put seconds there all queries are completed with less then 5ms which dives us totally flat CDF. I think average query processing time should be around 10ms for us so we can ether divide measurement by 100 which would generate better CDF but in weird units of measure (10ms) or we can use 'ms' as a unit of measure, but use custom histograms with bins upscaled by 10. IMO second option is better.

Reply:
I've played with different bin sizes and I have to say that it's hard to define meaningful ranges for both debug and release builds. This is what I get for `cargo build --release` + `pgbench -c4 -j4 -s10 -T40`:

```
$ curl -s localhost:9898 | rg get_page_at_lsn
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.00001"} 2
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.0001"} 99353
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.00015"} 153803
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.0002"} 186736
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.00025"} 226120
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.0003"} 272001
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.00035"} 299481
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.0005"} 324567
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.00075"} 330217
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.001"} 331456
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.0025"} 333337
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.005"} 333436
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.0075"} 333456
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.01"} 333465
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.0125"} 333467
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.015"} 333469
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.025"} 333474
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.05"} 333475
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="0.1"} 333478
pageserver_smgr_query_time_bucket{smgr_query_type="get_page_at_lsn",le="+Inf"} 333481
pageserver_smgr_query_time_sum{smgr_query_type="get_page_at_lsn"} 68.62459031499849
pageserver_smgr_query_time_count{smgr_query_type="get_page_at_lsn"} 333481
```

```
$ curl -s localhost:9898 | rg get_rel_exists
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.00001"} 116
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.0001"} 205
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.00015"} 220
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.0002"} 226
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.00025"} 227
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.0003"} 227
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.00035"} 228
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.0005"} 229
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.00075"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.001"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.0025"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.005"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.0075"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.01"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.0125"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.015"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.025"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.05"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="0.1"} 230
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_exists",le="+Inf"} 230
pageserver_smgr_query_time_sum{smgr_query_type="get_rel_exists"} 0.007725188999999994
pageserver_smgr_query_time_count{smgr_query_type="get_rel_exists"} 230
```

```
$ curl -s localhost:9898 | rg get_rel_size
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.00001"} 1249124
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.0001"} 1286018
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.00015"} 1289263
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.0002"} 1289998
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.00025"} 1290208
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.0003"} 1290273
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.00035"} 1290312
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.0005"} 1290363
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.00075"} 1290392
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.001"} 1290402
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.0025"} 1290406
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.005"} 1290409
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.0075"} 1290413
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.01"} 1290442
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.0125"} 1290444
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.015"} 1290445
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.025"} 1290448
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.05"} 1290449
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="0.1"} 1290449
pageserver_smgr_query_time_bucket{smgr_query_type="get_rel_size",le="+Inf"} 1290449
pageserver_smgr_query_time_sum{smgr_query_type="get_rel_size"} 4.1290146209995395
pageserver_smgr_query_time_count{smgr_query_type="get_rel_size"} 1290449
```

Maybe we should use different bins for each smgr query type.